### ISSUE TITLE
Bible Memory Feature

### ISSUE THREAD

**Issue by ghost** [AndBible/and-bible#2547]
Hello!

Thank you for a wonderful tool!. I love the app.. recently I was wanting to try and memorize scripture more and while I found several tools that did a good job they weren't really open source or Foss which I have come to appreciate and look for in an app. Its also yet another app to install. I thought it might make sense to be incorporated into the app I already use for studying Gods word and thus this feature request..

I have been using the play store Verse Locker app by scripturememory.com.. I like the options it has for memorizing.. like fill in the blank, listen etc.. I feel like the cloud login and network component would be unnecessary if it were built into andbible so maybe there isnt to much extra stuff to add since the scripture is already present . I realize this might not be in scope but I figured I would put it out there. 

Thank you again!

**Comment by tuomas2** [comment]
Related #1356

**Comment by agrogers** [comment]
It's a nice idea but it's quite a lot of work to do well and probably to
the side of what the app is intended for. Thanks for the suggestion though.

___________________________
*www.TenOutOfTen.org* <https://www.TenOutOfTen.org>
***@***.***
(+61) 0477 174 193 (Australia)
(+95) 09 250018669 (Myanmar)

**Comment by siesta5787** [comment]
I would argue bible memory is part of studying and using the bible so very related :) having said that any estimation on the hours required to incorporate something like this? Since I see this has the sponsor me tag on it? 

Personally I have used several of the other apps and even recently used this one https://www.remem.me which is open source though its not my favorite.  Most require registration and separate apps and importing the verses manually. Maybe you could use some of the code for the "memory games" from remem me to save time? Not sure if licensing is right for that or not but figured I would mention it.

**Comment by tuomas2** [comment]
I checked VerseLocker [demo video](https://scripturememory.com/products-resources/digital-tools/mobile-apps), here are rough estimates:

Basic framework to make this in AndBible: 1-2 hours.
Then it needs some functionality: There were 4 features:
- Word blur
- Word scramble
- Initials only: difficult to implement as "tooltip" is not possible without a mouse, but perhaps full word would be shown by tapping the initial
- Type it

I think those are all doable with about 2-3 hours of work each. So minimum (with one feature) 3-5 hours. All features maybe one full working day or a bit more.

**Comment by siesta5787** [comment]
OK cool. I would be willing to start the thing with 2 hours of contribution if maybe we could get a few more people who were interested to also contribute (better way notify people of this?). 

Would the basic functionality include like a place to see like a flashcard view of the verse, with categories of in progress or memorized and a way to add new verses?  Would be nice to select a verse or passage in the bible and say "add to memory" and have it create flashcard in the memory section.  

I would think a good first feature would be blur option

Maybe once the base in there people will contribute for other features they want.  

Thanks!

**Comment by tuomas2** [comment]
"Basic framework" does not include any functionality yet. "Minimum viable product" will be 3-5 hours.

**Comment by meichthys** [comment]
I've also been using https://biblememory.com, but an open-source implementation for AndBible is a great Idea!

**Comment by meichthys** [comment]
@siesta5787 are you still willing to support part of this? If so, I'll match your contribution to support the Minimum viable product.

**Comment by siesta5787** [comment]
@meichthys hi yes I would do that! I will do 2 hrs if you match I think that will give us a minimum viable product.  Then hopefully we can get a few more people to contribute to build out other features.

**Comment by meichthys** [comment]
Great to hear. Done on my end!

**Comment by siesta5787** [comment]
@meichthys @tuomas2  I just did as well. Forgot to make a comment to say what it was for but just did it.

**Comment by tuomas2** [comment]
Thank you @siesta5787  and @meichthys! I am now prioritizing this issue based received sponsorship. 
I have some still some important maintenance tasks pending and waiting in queue next but will do this as soon as possible otherwise. Probably within a month of two.

**Comment by tuomas2** [comment]
Implementation plan
- User selects one or more verses that he wants to memorize
- User selects "Memorize" function from modal that pops up (the same modal that shows bookmark/note/share/compare... functions). 
- "Special document" will open (similarly to "compare" function), that has the actual memorizing functionality.

**Comment by tuomas2** [comment]
Test build for memorize feature coming soon: https://github.com/AndBible/and-bible/releases/tag/build-870-memorize

**Comment by tuomas2** [comment]
Word blur: 

![Image](https://github.com/user-attachments/assets/650d901a-8790-4366-a903-fb9f0758bbb3)

Word scramble:

![Image](https://github.com/user-attachments/assets/2b48d551-37a3-44c6-b350-abc2634f18bb)

UI not very finished looking yet.

**Comment by meichthys** [comment]
Great progress!

**Comment by agrogers** [comment]
Nice. Well done. Were you recording the number of attempts? Or better the
attempt and it's quality? Maybe 0 for dont know, 1 for know and 2 for know
it well. That could form the basis for a simple sorting to focus on the
verses i still struggle to remember (a small step towards spaced
repetition).

___________________________
*www.TenOutOfTen.org* <https://www.TenOutOfTen.org>
***@***.***
(+61) 0477 174 193 (Australia)
(+95) 09 250018669 (Myanmar)

**Comment by tuomas2** [comment]
> Nice. Well done. Were you recording the number of attempts? Or better the
> attempt and it's quality? Maybe 0 for dont know, 1 for know and 2 for know
> it well. That could form the basis for a simple sorting to focus on the
> verses i still struggle to remember (a small step towards spaced
> repetition).
> […](#)

Good idea. Can create a follow-up feature request about that. I am planning to wrap this up after with a bit cleaner UI, not adding more features for now. Already went beyond MVP with 2 modes :-P

**Comment by siesta5787** [comment]
I like it! Thanks. Great base for this feature. Hopefully other might contribute a little more in the future to finish of.  I think some from of tracking or list view would be nice. Like check off "memorized " or in progress" and it goes to a note or list section of sort where you can view all you are working on and work thru them from there for practice.

**Comment by meichthys** [comment]
Yes, some sort of tracking would be nice. My thought is to have two separate bookmarks, one called 'Memorize' and the other called 'Memorized'. Then from the memory screen, the user could choose any of the verses currently bookmarked with 'Memorize'. Once the user successfully memorizes the verse, the bookmark for that verse could be changed to 'Memorized'. It would then be easy to show some basic stats of how many verses are memorized and how many the user wants to memorize.

Another idea I thought might be neat would be to visually show which verses have been memorized in the verse chooser. I was thinking that the books and chapter squares could have a subtle "progress bar" on the top of them showing the percentage of the book and the percentage of the chapter that have been memorized. Something like this (If I have memorized 3John 1:1-6):

![img1.jpg](https://github.com/user-attachments/assets/63d2356e-f6e6-49da-a151-71e2949bdd10)

**Comment by meichthys** [comment]
Yes, some sort of tracking would be nice. My thought is to have two separate bookmarks, one called 'Memorize' and the other called 'Memorized'. Then from the memory screen, the user could choose any of the verses currently bookmarked with 'Memorize'. Once the user successfully memorizes the verse, the bookmark for that verse could be changed to 'Memorized'. It would then be easy to show some basic stats of how many verses are memorized and how many the user wants to memorize.

Another idea I thought might be neat would be to visually show which verses have been memorized in the verse chooser. I was thinking that the books and chapter squares could have a subtle outline (or background) around them showing the percentage of the book and the percentage of the chapter that have been memorized. The verse squares would just show an outline around the verses that were memorized and no outline for the verses that were not memorized. Something like this (If I have memorized 1John 1:1-6):

![img1.jpg](https://github.com/user-attachments/assets/da3787b9-4c4c-428d-84bd-9679bbb2dec4)

**Comment by tuomas2** [comment]
New test build coming soon: https://github.com/AndBible/and-bible/releases/tag/build-870-memorize2

Some bug fixes and a little ui tweaks.

**Comment by siesta5787** [comment]
@meichthys I like those ideas!

**Comment by tuomas2** [comment]
Merged feature now, hours used so far 13h.
My original max estimate was 8h with 2 features, so exceeded my original estimate by 5 hours. 
But well, business as usual in software development....

**Comment by meichthys** [comment]
Easy for me to say, but your work on this will not to be without eternal value. 👍
